rtsp_to_hls/server.exs

```diff
@@ -0,0 +1,239 @@
+    request_type = do_serve(socket, state, [:describe, :get_parameter, :setup], [:play])
+    new_state = if request_type == :setup, do: :ready, else: :init
+    serve(socket, %{state | cseq: state.cseq + 1, server_state: new_state})
+  end
+
+  defp serve(socket, state) when state.server_state == :ready do
+    request_type = do_serve(socket, state, [:describe, :get_parameter, :setup, :play])
+
+    new_state =
+      if request_type == :play do
+        if not is_nil(state.stream_ctx) do
+          pipeline_opts = %{
+            client_ip: state.ip,
+            client_port: state.client_port,
+            ssrc: @test_ssrc,
+            pt: @test_pt,
+            clock_rate: @test_clock_rate,
+            server_udp_port: @test_udp_port
+          }
+
+          Membrane.Pipeline.start_link(Pipeline, Map.merge(pipeline_opts, state.stream_ctx))
+        end
+
+        :playing
+      else
+        :ready
+      end
+
+    serve(socket, %{state | cseq: state.cseq + 1, server_state: new_state})
+  end
+
+  defp serve(socket, state) when state.server_state == :playing do
+    if is_nil(state.stream_ctx) do
+      # For signalling test, don't respond to GET_PARAMETER keep-alives in playing state
+      do_serve(socket, state, [:describe, :setup, :play])
+    else
+      do_serve(socket, state, [:describe, :get_parameter, :setup, :play])
+    end
+
+    serve(socket, %{state | cseq: state.cseq + 1})
+  end
+
+  defp do_serve(socket, state, respond_on, raise_on \\ []) do
+    request = get_request(socket)
+
+    request_type =
+      case request do
+        "DESCRIBE " <> _rest -> :describe
+        "SETUP " <> _rest -> :setup
+        "PLAY " <> _rest -> :play
+        "GET_PARAMETER " <> _rest -> :get_parameter
+        _other -> raise("RTSP Endpoint sent unrecognised request: #{inspect(request)}")
+      end
+
+    response_body =
+      if Enum.any?(respond_on, fn allowed_request -> request_type == allowed_request end) do
+        generate_response(request_type, state)
+      end
+
+    if is_nil(response_body) do
+      if Enum.any?(raise_on, fn erroneous_request -> request_type == erroneous_request end) do
+        raise("""
+        RTSP Endpoint sent invalid request: #{inspect(request)}
+        to fake server in state: #{inspect(state.server_state)}
+        """)
+      end
+    else
+      :gen_tcp.send(socket, "RTSP/1.0 200 OK\r\nCSeq: #{state.cseq}\r\n" <> response_body)
+    end
+
+    request_type
+  end
+
+  defp get_request(socket, request \\ "") do
+    case :gen_tcp.recv(socket, 0) do
+      {:ok, packet} ->
+        request = request <> packet
+        if packet != "\r\n", do: get_request(socket, request), else: request
+
+      {:error, :closed} ->
+        exit(:normal)
+
+      {:error, reason} ->
+        raise("Error when getting request: #{inspect(reason)}")
+    end
+  end
+
+  defp generate_response(request_type, state) do
+    case request_type do
+    :describe ->
```

Comment:
seems like this is not formatted

Reply:
indeed, I have added these two script files to the formatter config